# http://satijalab.org/seurat/pbmc3k_tutorial.html
library(dplyr)
library(Matrix)
library(scales)
library(Seurat)
library(RColorBrewer)

# Import data
lbl.data <- lbl.data # separate label from numeric data
num.data <- all.data



# Obtain top most varied genes
seu.data <- CreateSeuratObject(num.data, project = "single cell analysis")
seu.data <- AddMetaData(seu.data, metadata = lbl.data)
seu.data <- NormalizeData(seu.data)
seu.data <- ScaleData(seu.data, model.use = "negbinom")


seu.data <- FindVariableGenes(seu.data)
seu.data <- RunPCA(seu.data, pcs.compute = 30, weight.by.var = F , do.print = F) 
PCElbowPlot(seu.data)

seu.data <- FindClusters(seu.data, reduction.type = "pca", dims.use = 1:10, 
                         resolution = 0.2, print.output = 0, save.SNN = F) 

PCAPlot(seu.data,1,2,pt.size=0.6,group.by = "Genotype")
# Finds clusters using the PCA


seu.data <- RunTSNE(seu.data, dims.use = 1:10, do.fast = T)
seu.data <- DBClustDimension(seu.data,1,2,reduction.use="tsne",G.use=4)

DotPlot(seu.data,genes.plot=c("GCG","INS","SST","PPY","KRT19"))
VlnPlot(seu.data,group.by="Genotype",features.plot="GCG",point.size.use=0)


TSNEPlot(seu.data, do.label = F, no.legend = F, pt.size = 0.6) 
TSNEPlot(seu.data, group.by = "Genotype", pt.size = 0.6)
TSNEPlot(seu.data, group.by = "Batch", pt.size = 0.6)


# Plot cell type markers
FeaturePlot(seu.data, features.plot = "GCG", cols.use = c("grey", "blue"), reduction.use = "tsne", no.legend = F, pt.size = 0.6) # alpha
FeaturePlot(seu.data, features.plot = "INS", cols.use = c("grey", "blue"), reduction.use = "tsne", no.legend = F, pt.size = 0.6) # beta
FeaturePlot(seu.data, features.plot = "HNF1A", cols.use = c("grey", "blue"), reduction.use = "pca", no.legend = F, pt.size = 0.6)

#### more analysis
#seu.data@cell.names
#seu.data@ident

cells <- rownames(seu.data@meta.data)[which(seu.data@meta.data$res.0.6==2)]
#cells_0 <- rownames(seu.data@meta.data)[which(seu.data@meta.data$res.0.6==0)]
#cells_6 <- rownames(seu.data@meta.data)[which(seu.data@meta.data$res.0.6==6)]
seu.data.sub <- SubsetData(object = seu.data,cells.use = cells)
cluster0.markers <- FindMarkers(object = seu.data, ident.1 = 0, min.pct = 0.25,thresh.test=2)

### markers generated by findmarker funciton,
#markers <- cluster0.markers[which(cluster0.markers$p_val_adj<1e-4),]
#markers <- markers[,c(1,2,5)]
#write.table(markers,"markers_for_beta.txt",quote = F,sep="\t",col.names = NA)
markers <- read.table("beta_markers.txt", sep="\t", header = F, stringsAsFactors = F)[,1]

lbl.data <- lbl.data
lbl.data.sub <- lbl.data[colnames(seu.data.sub.0),]

seu.data.sub <- CreateSeuratObject(seu.data.sub.0, project = "Subset analysis")
seu.data.sub <- AddMetaData(seu.data.sub, metadata = lbl.data.sub)

seu.data.sub <- NormalizeData(seu.data.sub)
seu.data.sub <- ScaleData(seu.data.sub,model.use = "negbinom")

#seu.data.sub <- FindVariableGenes(seu.data.sub)
seu.data.sub <- RunPCA(seu.data.sub, pc.genes = markers, pcs.compute = 30, weight.by.var = F)

PCAPlot(seu.data.sub,1,2,pt.size=0.8,group.by = "Genotype")
PCElbowPlot(seu.data.sub)

###
seu.data.sub <- RunTSNE(seu.data.sub, dims.use = 1:6, do.fast = F,perplexity=30)


TSNEPlot(seu.data.sub, group.by = "Genotype", pt.size = 1.0)
VlnPlot(seu.data.sub,group.by="Genotype",features.plot="TTR",point.size.use=0)
VlnPlot(seu.data.sub,group.by="Genotype",features.plot="INS",point.size.use=0)


FeaturePlot(seu.data.sub, features.plot = "GCG", cols.use = c("grey", "blue"), reduction.use = "tsne", no.legend = F, pt.size = 1.0) # alpha
FeaturePlot(seu.data.sub, features.plot = "INS", cols.use = c("grey", "blue"), reduction.use = "tsne", no.legend = F, pt.size = 1.0) # beta

save(seu.data.sub,file="seu.data.sub.rda")

###scatter plot 
seu.norm <- seu.data.sub@data
seu.norm.red <- seu.norm[,which(lbl.data.sub$Genotype=="KO2")]
seu.norm.blue <- seu.norm[,which(lbl.data.sub$Genotype=="WT")]
seu.norm.green <- seu.norm[,which(lbl.data.sub$Genotype=="R200Q")]

pdf("scatter.pdf")
plot(seu.norm["GCG",],seu.norm["INS",],cex=0.5,pch=20,xlab="GCG_expr_logtpm",ylab = "INS_expr_logtpm",col="white")
points(seu.norm.red["GCG",],seu.norm.red["INS",],col="red",cex=0.5,pch=20)
points(seu.norm.blue["GCG",],seu.norm.blue["INS",],col="blue",cex=0.5,pch=20)
points(seu.norm.green["GCG",],seu.norm.green["INS",],col="green",cex=0.5,pch=20)
legend("bottomleft", legend=c("KO2", "WT","R200Q"),fill=c("red", "blue","green"), cex=1.0)
dev.off()


### differential expressed genes
seu.data.sub@ident <- lbl.data.sub$Genotype
des_wt <- FindMarkers(object = seu.data, ident.1 = "WT", ident.2=c("KO2","R200Q"),genes.use=id_sub$SYMBOL,
                      min.pct = 0.25,logfc.threshold = 0.4,thresh.test=1)

markers <- des_wt[which(des_wt$p_val_adj<1e-4),]
markers <- markers[,c(1,2,5)]
markers <- markers[order(markers$avg_logFC,decreasing = T),]
write.table(markers,"des_for_alpha_and_beta.txt",quote = F,sep="\t",col.names = NA)

markers.0 <- read.table("des_for_beta.txt", sep="\t", row.names=1, header = T, stringsAsFactors = F)
markers <- read.table("des_for_alpha_and_beta.txt", sep="\t", row.names=1, header = T, stringsAsFactors = F)

#inter_markers <- intersect(rownames(markers),rownames(markers.0))


######for the new figures
library(Seurat)
lbl.data <- lbl.data
lbl.data.sub <- lbl.data[colnames(seu.data.sub.0),]

seu.data.sub <- CreateSeuratObject(seu.data.sub.0, project = "Subset analysis")
seu.data.sub <- AddMetaData(seu.data.sub, metadata = lbl.data.sub)
seu.data.sub@ident <- lbl.data.sub$Genotype

seu.data.sub <- NormalizeData(seu.data.sub)
seu.data.sub <- ScaleData(seu.data.sub,model.use = "negbinom")

cluster0_scale <- seu.data.sub@scale.data
cluster0_scale <- cluster0_scale[,which(cluster0_scale["INS",]>0)]
#cluster0_scale <- colnames(seu.data.sub@scale.data[,which(seu.data.sub@scale.data["INS",]>0)])

#cluster2_scale_alpha <- cluster2_scale[,which(cluster2_scale["INS",]<0)]
#cluster2_scale_imb <- cluster2_scale[,which(cluster2_scale["INS",]>0)]

lbl.data.mb <- lbl.data[colnames(cluster0_scale),]
cells <- colnames(cluster0_scale)
seu.data.sub <- SubsetData(object = seu.data.sub,cells.use = cells)

#VlnPlot(seu.data.sub,group.by="Genotype",features.plot="GCG",point.size.use=0)
VlnPlot(seu.data.sub,group.by="Genotype",features.plot="INS",point.size.use=0)


### differential expressed genes again

deg_wt <- FindMarkers(object = seu.data.sub, ident.1 = "WT", ident.2=c("KO2","R200Q"),
                           min.pct = 0.25,logfc.threshold = 0.35,thresh.test=1)
deg_wt_2 <- FindMarkers(object = seu.data.sub, ident.1 = "WT", ident.2="R200Q",
                        min.pct = 0.25,logfc.threshold = 0.35,thresh.test=1)

markers <- deg_wt_2[which(deg_wt_2$p_val_adj<1e-4),]
markers <- markers[,c(1,2,5)]
markers <- markers[order(markers$avg_logFC,decreasing = T),]
write.table(markers,"des_for_maturebeta_wtvsr200q.txt",quote = F,sep="\t",col.names = NA)

pdf("heatmap.0.maturebeta_wtvsr200q.pdf",height = 12,width=10)
DoHeatmap(seu.data.sub,genes.use = rownames(markers),slim.col.label = T,remove.key = T)
dev.off()

lbl.data.syp <- lbl.data[colnames(data.scale[,which(data.scale["SYP",]>0)]),]
lbl.data.pdx <- lbl.data[colnames(data.scale[,which(data.scale["PDX1",]>0)]),]



#### enriched pathway analysis
pathways <- read.table("pathways.txt", sep="\t", header = F, stringsAsFactors = F)
pathways[,1] <- sapply(strsplit(pathways[,1], split='_', fixed=T), function(x) (x[1]))
colnames(pathways) <- c("Pathway","Gene","Cell_type")


#rl <- list()
rl <- c(pathways[1,1],unlist(strsplit(pathways$Gene[1], split=';', fixed=T))[1],pathways[1,3])
for (i in 1:19){
 gl <- unlist(strsplit(pathways$Gene[i], split=';', fixed=T))
 for (j in 1:length(gl)){
  a <- c(pathways[i,1],gl[j],pathways[i,3])
  rl <- rbind(rl,a)
 }
}
colnames(rl) <- c("Pathway","Gene","Cell_type")
rl <- as.data.frame(rl[-1,])


des_alpha <- read.table("des_for_maturealpha_wtvsothers.txt", sep="\t",row.names = 1, header = T, stringsAsFactors = F)
rl_alpha <- rl[which(rl$Cell_type=="Mature alpha"),]
des_alpha <- des_alpha[as.character(rl_alpha$Gene),]
rl_alpha$logFC <- -(des_alpha$avg_logFC)


library(ggplot2)
pdf("alpha_pathway.pdf")
ggplot(rl_alpha, aes(x=Gene, y=logFC, fill=Pathway)) + 
  geom_bar(stat="identity") +
  coord_flip()
dev.off()
